### PR TITLE
Handle Half-open TCP sockets

### PR DESCRIPTION
**Pull request by rianhunter**
If the host machine that the Dropbox client SDK is running on misses a
FIN packet while it's waiting for a response from the server it will
hang forever by default. This can happen if an intermediate router
transiently goes down, or if the host is suspended via ACPI power
management. This problem is common and widespread and documented here:

https://en.wikipedia.org/wiki/TCP_half-open#RFC_793

A thorough description of the problem can be found here:

http://blog.stephencleary.com/2009/05/detection-of-half-open-dropped.html

This change makes all HTTP API requests instead use a default timeout
of 30 seconds instead of unconditionally blocking forever.  It adds a
special case for files_list_folder_longpoll which may have a
user-defined timeout.

### CODE REVIEW

**Comment by rianhunter** [comment]
Note that this timeout does not represent the maximum amount of time allowed for downloading the entire response, it's the maximum amount of time allowed to wait on receiving any data during the download.

**Comment by rianhunter** [comment]
Ping :)

**Comment by braincore** [comment]
Thanks for this! I'm thinking we should take an approach closer to #54. Can you have the `Dropbox` constructor take an optional timeout? If it's unset, then we can set it to the `socket.getdefaulttimeout()`, otherwise, use the passed in value.

Unless you think there's a need to customize the timeout on a per request basis, but based on your description (and my awareness of the problem), I don't think that's necessary.

**Comment by braincore** [comment]
Oh, sorry, of course you have it on a per request basis for the longpoll route. Unfortunately, `base.py` is auto-generated, so we can't simply add +100 there. Can you take the approach I said above, but in the final request method `request_json_string()` check for whether the target host is `_HOST_NOTIFY`? If so, artificially bump the timeout to 100.

**Comment by braincore** [comment]
Let me know if you just want me to hammer this out as I can do it quickly.

**Comment by rianhunter** [comment]
Yeah I agree with your feedback.. I will change, thanks!

One piece of feedback for your end, a generically short socket timeout
(e.g. the de facto 30 second timeout) may cut short the server-side
timeout for the longpoll method.  It would be nice if the client could
tell the notification server to jitter within an upper bound instead
of the server always deciding to randomly jitter for 90 seconds after
the user specified timeout. i.e.:

longpoll(timeout=30) => times out between 0-30 seconds

instead of

longpoll(timeout=30) => times out between 30-90 seconds (current behavior)

This would help the server jitter more effectively for clients who
always decide to hard timeout at shorter intervals.

**Comment by rianhunter** [comment]
Okay I updated the diff.

I don't use `socket.getdefaulttimeout()` because that doesn't seem appropriate at this layer.

If the user does nothing, they will use a timeout value of 30 seconds which will prevent hangs (except for longpoll which will use a timeout based on its argument). They can revert back to the old behavior by passing timeout=None to the Dropbox constructor.

File: dropbox/dropbox.py [hunk 1]
```diff
@@ -185,11 +189,14 @@ def __init__(self,
...
 
+        self._timeout = timeout
+
     def request(self,
                 route,
                 namespace,
                 request_arg,
-                request_binary):
+                request_binary,
+                timeout=None):
```
Comment (braincore) [inline dropbox/dropbox.py]: Please document the new parameter in the docstring. Here and in `request_json_object()`.

File: dropbox/dropbox.py [hunk 2]
```diff
@@ -211,11 +218,25 @@ def request(self,
         route_style = route.attrs['style'] or 'rpc'
         serialized_arg = stone_serializers.json_encode(route.arg_type,
                                                        request_arg)
+
+
+        if (timeout is None and
+            route == files.list_folder_longpoll):
```
Comment (braincore) [inline dropbox/dropbox.py]: One more indent on this line continuation. Also, good way to do it (specifically checking the route)!

**Comment by braincore** [comment]
Small nits and then I'll rebase it is. Thanks!

**Comment by rianhunter** [comment]
Okay added indent and documentation.

**Comment by braincore** [comment]
Merged in 397c2e1c332b971a0ce7601bb6d2674ccd392528.

**Comment by rianhunter** [comment]
Can you re-pull from my repo? made some important documentation improvements.

**Comment by ijunaid8989** [comment]
is there any way to pass timeout none to HTTP API?

**Comment by greg-db** [comment]
@ijunaid8989 This is an old pull request. If you're having trouble with the Dropbox Python SDK or have a question about how to accomplish something, please [open a new issue](https://github.com/dropbox/dropbox-sdk-python/issues/new) with more information so we can help.

**Comment by ijunaid8989** [comment]
I am not using python sdk. I am using an Elixir Client and sending requests using API. And I am having timeout on upload session start . I want to know how I can set timeout to none in direct HTTP API request

**Comment by greg-db** [comment]
@ijunaid8989 This is a pull request for the official Dropbox Python SDK. We can't provide support for Elixir Client, e.g., for information on how to set or disable a timeout for that client, as that's not made by Dropbox. You may need to refer to the documentation or developer for the client for help.

If you're having trouble with the Dropbox API itself, you can [share the details in a post on the forum](https://www.dropboxforum.com/t5/forums/postpage/choose-node/true/board-id/101000014) or [open a ticket](https://www.dropbox.com/developers/contact).

**Pull request closed without merging**